[SYCL] Don't select devices with no available images

npmiller opened the pull request:
This patch solves:
* https://github.com/intel/llvm/issues/2004

And is related to:
* https://github.com/intel/llvm/issues/1665

In some cases the current selector may select a device for which we don't have an AOT or SPIR-V binary for, this patch ensures that such devices get skipped.

cperkinsintel commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -1450,6 +1450,26 @@ kernel_id ProgramManager::getSYCLKernelID(const std::string &KernelName) {
   return KernelID->second;
 }
 
+bool ProgramManager::hasCompatibleImage(const device &Dev) {
```
maybe instead of running two loops and creating a temporary set, would it be possible to use `std::find` or `std::any_of` on the DeviceImages collection ? That might be easier to read and quicker.

npmiller replied:
Updated with a slightly easier to use map and use of `std::any_of`

npmiller commented:
So after having another look at this I think it's actually incorrect to reject a device if it doesn't have an image, first off as the lit tests show it breaks `sycl-ls` since that uses device selectors but obviously doesn't have any device images.

I'll re-work this patch to take availability of device image into account in the device scores instead of just rejecting the device.

npmiller commented:
So this turned into a bit of a larger refactoring of the device selector, here's the main changes:

* Removed the initial score of `1000` when `SYCL_DEVICE_FILTER` is set. This is unnecessary as `SYCL_DEVICE_FILTER` filters the list of devices itself, so we can just use regular scoring on the already filtered list of devices
* Replaced `isDeviceOfPreferredSyclBe` with a new `getDevicePreference`, which returns a "preference" score to add to the device score, this preference score includes preference of the level_zero backend, as well as a strong preference  for devices that have images in the binary.
* Remove use of `isDeviceOfPreferredSyclBe` as tie-breaker in `select_device`, this was redundant because it was already used to boost the score of preferred devices, so it should be included in the score.

joeatodd commented on sycl/source/device_selector.cpp:
```diff
@@ -64,11 +77,12 @@ device device_selector::select_device() const {
     if ((score < dev_score) ||
-        (score == dev_score && isDeviceOfPreferredSyclBe(dev))) {
+        ((score == dev_score) &&
+         (getDevicePreference(dev) > getDevicePreference(*res)))) {
```
Totally trivial nitpick: 
```suggestion
         (getDevicePreference(*res) < getDevicePreference(dev)))) {
```
slightly improves the readability of the logic, as the order matches the first two comparisons.

cperkinsintel reviewed (approved):
LGTM.

steffenlarsen commented on sycl/source/device_selector.cpp:
```diff
@@ -97,23 +111,10 @@ device device_selector::select_device() const {
-  // device::get_devices returns filtered list of devices.
-  // Keep 1000 for default score when filters were applied.
-  if (FilterList)
-    Score = 1000;
```
Why is this removed? I am not completely sure, but it seems like this tries to ensure that no device is rejected if the user applied a filter. Maybe we could simply change it to a `+= 1` to ensure this?

bso-intel reviewed (changes requested):
default_selector can throw an exception if no device is available to satisfy the SYCL_DEVICE_FILTER.
https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md

npmiller commented:
Yes and that still happens, `SYCL_DEVICE_FILTER` affects the list of devices, whereas the device selectors only score and rank the available devices, with the possibility of rejecting them.

If there is no device that satisfies `SYCL_DEVICE_FILTER`, the list of devices will be empty to begin with, so the default_selector won't have anything to score, and it will correctly throw a device not found exception.

steffenlarsen reviewed (approved):
Changes LGTM. @bso-intel approval required.

bso-intel reviewed (approved):
It makes sense. If SYCL_DEVICE_FILTER is set, no device that does not satisfy the filter can be created.
So, no additional score for filter is needed.
LGTM.
